Add swr caching for selected account

lujakob opened the pull request:
### Describe the changes you have made in this PR

Replace custom caching code in lib/api and lib/cache to get account info with SWR(https://github.com/vercel/swr)

I was interested in this issue and started debugging and testing. I got it working at some point, that's why i share it here. I'm not sure if this is how it was intended however. Feel free go give me any suggestions on how to modify or improve. In case it doesn't help at all, I'm not bothered by discarding the whole PR in favour of other implementations.

### Link this PR to an issue [optional]

Fixes #1923

### Type of change

- `feat`: New feature (non-breaking change which adds functionality)

### How has this been tested?

Manually so far

### Checklist

- [ ] My code follows the style guidelines of this project and performed a self-review of my own code
- [ ] New and existing tests pass locally with my changes
- [ ] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

lujakob commented:
@rolznz I will probably clean this up and might change again tomorrow. Please have a look and let me know what you think. Thanks.

rolznz commented on src/common/lib/api.ts:
```diff
@@ -33,64 +29,76 @@ export interface AccountInfoRes {
+export const useAccountInfoCached = async (
+  id: string | null
+): Promise<AccountInfo | null> => {
+  const { data, error } = await useSWR<AccountInfoRes>(
```
can you return the mutate function here too? unless you need to mutate an account that is not currently selected?

lujakob replied:
Yes, good point. By refactoring the `useSWR` to the component level I can easily use bound mutate. This really simplifies things.

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -32,11 +26,9 @@ interface AccountContextType {
-  fetchAccountInfo: (options?: {
-    accountId?: string;
-  }) => Promise<AccountInfo | undefined>;
+  updateAccountInfo: () => Promise<void>;
```
I wonder if this method could be renamed to fetchAccountInfo? in my opinion it doesn't actually "update" the account

rolznz commented on src/common/lib/api.ts:
```diff
@@ -143,7 +153,8 @@ export default {
   swr: {
-    getAccountInfo: swrGetAccountInfo,
+    useAccountInfoCached: useAccountInfoCached,
+    mutateAccountInfo: mutateAccountInfo,
```
I don't think these methods should live in an "swr" object (I know it was originally like this, I don't think it's the right approach). We can use swr with any of the api methods. Also, I don't think we should introduce mutation functions here that didn't exist, especially if they only mutate the cached SWR data.

lujakob replied:
By introducing swr we split the functionality of initial fetching on account change (in AccountContext) and  updating/mutating (in any other component that triggers an update/mutation). Hence there need to be two functions.
I will remove the "swr" nesting, to me it also seemed strange.

rolznz replied:
another option is that you can use the mutate function from the useSWR hook in the component that needs to update it. I think we should have a separation from swr and the API methods. SWR is just an optional way to retrieve data, it shouldn't be mixed in with actual API methods.

rolznz commented on src/common/lib/api.ts:
```diff
@@ -33,64 +29,76 @@ export interface AccountInfoRes {
+};
+
+const buildAccountInfo = (
+  id: string | null,
```
the id should exist, I don't think it should be nullable here. I think we should never use buildAccountInfo if the id or data isn't available.

lujakob replied:
This is actually due to a Typescript type problem: due to the id coming from the calling function "useAccountInfoCached" that calls the "useSWR" hook. By design the hooks need to be called every single render turn and the mechanics to skip it is to pass a key of null to useSWR. That's why the parent function id parameter can be null and the type here therefore as well. Do you have an idea how to solve it? If I try non-null assertion "id!" with the id argument, I get the lint error: "Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion"

lujakob replied:
Actually I'm using "// eslint-disable-next-line @typescript-eslint/no-non-null-assertion" in this function. I can move it up to the calling function. But the issue of an id parameter of null coming in will stay there, because of what I explained above. Unless you have a better idea.

rolznz commented on src/common/lib/api.ts:
```diff
@@ -33,64 +29,76 @@ export interface AccountInfoRes {
+export const mutateAccountInfo = async (
+  id: string | null
+): Promise<AccountInfo | null> => {
+  const data = await mutate<AccountInfoRes>(
```
Just mutating the value should be enough - SWR will return the new value in the next render. You don't need to return buildAccountInfo() as the response.

lujakob replied:
If I remove this the AccountMenu balance of the selcted account will not be updated after having a payment success of the "Receive" page. As far as I see it, there is no render in the AccountContext when the screen of the Receive page has received the account. For this reason the mutate being called, needs to refetch the account data and update the states in AccountContext

`
  const refetchAccountInfo = async () => {
    const accountInfo = await api.refetchAccountInfo(account?.id || null);
    if (accountInfo) {
      // eslint-disable-next-line
      setAccount(accountInfo as any);
      updateAccountBalance(accountInfo.balance, accountInfo.currency);
    }
  };
`

lujakob replied:
It works now. This is resolved.

rolznz commented on src/common/lib/api.ts:
```diff
@@ -111,9 +119,11 @@ export const setSetting = (setting: MessageSettingsSet["args"]["setting"]) =>
     msg.request("removeAccount", { id }),
-    removeAccountFromCache(id),
+    // @Todo: how to remove account from swr cache?
+    // @Todo swr dows say "mutate(key)" will mark the data as expired and trigger a refetch for the resource
```
I don't think this is really necessary to delete it from the SWR cache since it's only stored for the duration of the session

rolznz commented:
> @rolznz I will probably clean this up and might change again tomorrow. Please have a look and let me know what you think. Thanks.

Thanks for your changes @lujakob !

I think some parts of this PR are more complicated than necessary. I hope my comments help.

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -93,46 +96,50 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
-    const id = options?.accountId || account?.id;
-    if (!id) return;
+  // hook: will be called each round, pass "id = null" to skip
+  const fetchAccountInfo = async () => {
```
This doesn't seem right. SWR hooks should be used within components or other hooks.

You can use an effect when the data from the SWR hook changes, to update the account info and balance, etc.

lujakob replied:
I refactored to call `useSWR` to the component level. This really simplifies things.

lujakob commented on src/app/context/AccountContext.tsx:
```diff
@@ -32,15 +33,37 @@ interface AccountContextType {
 
 const AccountContext = createContext({} as AccountContextType);
 
+export const getAccountInfoKey = (
```
@rolznz `getAccountInfoKey` and `buildAccountInfo` are also used in TestConnection. Should I create a `app/utils/accounts.ts` function and move them there?

rolznz replied:
I think it would be better to make a reusable hook e.g. src/app/hooks/useSelectedAccount and put all this logic in one place. I think it might just need an accountId parameter?

Then in the TestConnection and AccountContext you can just use the hook. This would also make AccountContext simpler too. What do you think?

lujakob replied:
I already used to have this abstracted as "useAccountInfoCached" in "api.ts" - then you mentioned to simplify things and call "useSWR" directly in the AccountContext. I guess it was a misunderstanding then and I also think it makes sense, so I refactor to abstract it and move it to the suggested place.

rolznz replied:
@lujakob sorry about that.

lujakob replied:
No worries.

lujakob commented on src/common/lib/api.ts:
```diff
@@ -33,64 +27,23 @@ export interface AccountInfoRes {
 }
 
-interface StatusRes {
+export interface StatusRes {
```
@rolznz I export the interfaces to be able to use them in AccountContext. Would it make sense to move them to "src/types.ts"?

rolznz replied:
That makes sense to me but let's leave them here for now to reduce the PR changes

lujakob commented:
@rolznz I addressed all of your comments and added three more questions for you. Do you have any further suggestions?

lujakob commented on src/app/screens/Options/TestConnection/index.tsx:
```diff
@@ -4,34 +4,39 @@ import Loading from "@components/Loading";
+  const [currentAccountId, setCurrentAccountId] = useState<string>("");
   const [errorMessage, setErrorMessage] = useState("");
   const [loading, setLoading] = useState(false);
+  const { data: accountInfoRes } = useSWR<AccountInfoRes>(
```
The TestConnection is actually a one time action after adding a new account. Does it even make sense to apply `useSWR` in this case? Wouldn't it even make more sense to just apply the fetcher function directly? 

`export const getAccountInfo = (): Promise<AccountInfoRes> =>
  msg.request<AccountInfoRes>("accountInfo")`

rolznz replied:
it will save the account in the SWR in memory cache for the next use (otherwise in this case it would have to fetch the account info twice), so it is nice to use SWR in my opinion

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -75,15 +111,15 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
-    [getFormattedFiat]
-  );
+  const refetchAccountInfo = async () => {
+    if (typeof mutate !== "function") return;
```
why is this needed? `if (typeof mutate !== "function") return;`

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -49,19 +72,32 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
   const [fiatBalance, setFiatBalance] = useState("");
+  const [, setStatus] = useState<StatusRes>();
+
+  const { data: accountInfoRes, mutate } = useSWR<AccountInfoRes>(
```
could you rename mutate? like `, mutate: mutateAccountInfo` so it is more obvious what it's doing?

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -75,15 +111,15 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
+  const refetchAccountInfo = async () => {
+    if (typeof mutate !== "function") return;
+    const data = await mutate();
+    setAccountData(account?.id, data);
```
I still don't think `setAccountData(account?.id, data);`  is needed. If you call mutate it will refetch the data using SWR and update the SWR cache.

lujakob replied:
Indeed. Now the AccountMenu is updated even though I don't setAccountData. Not sure why it didn't work when I tried initially. Good you persisted.

rolznz commented on src/app/context/AccountContext.tsx:
```diff
@@ -93,46 +129,45 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
 
+  useEffect(() => {
+    setAccountData(account?.id, accountInfoRes);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
can you fix the warning rather than disabling it?

lujakob replied:
I would need to add setAccountData and apply useCallback() - I didn't really get it working and seems like this makes it overly complicated. Is it a problem though? This exception is used in many other effects!?

rolznz replied:
It might work fine. It is just a code smell. I'm not sure how important it is to fix for now.

lujakob commented:
@rolznz I addressed all your concerns. I created the useSelectAccount hook and the AccountContext and TestConnection are even leaner now. Please have another look at it.

rolznz commented on src/app/hooks/useSelectedAccount.ts:
```diff
@@ -0,0 +1,49 @@
+};
+
+export function useSelectedAccount(id: string | undefined) {
+  const { data, mutate } = useSWR<AccountInfoRes>(
```
you can just do mutate: mutateAccountInfo and then the function will be memoized. Currently the way you're doing it will create a new mutateAccountInfo function each time the hook is accessed

rolznz commented on src/app/hooks/useSelectedAccount.ts:
```diff
@@ -0,0 +1,49 @@
+  };
+};
+
+export function useSelectedAccount(id: string | undefined) {
```
Sorry, I realized this should probably be called "useAccount" not "useSelectedAccount"

rolznz replied:
Edit: I see this would conflict with the AccountContext. hmm

lujakob replied:
I updated

rolznz commented on src/app/screens/Options/TestConnection/index.tsx:
```diff
@@ -68,6 +50,21 @@ export default function TestConnection() {
+      id: currentAccountId,
+    });
+    setAccountId(currentAccountId);
+    setCurrentAccountId(currentAccountId);
```
I don't understand this, why do we need another variable to track the current account? when we create a new account it should be selected, and we test that one. I don't think a user would change the account while the connection is being tested.

lujakob replied:
@rolznz what should happen when the user changes the selected account in the AccountMenu? My intention was to not change the screen data, because I thought the screen refers to the just before created Account. If you think it makes more sense for the screen info to reflect whatever account is or will be selected in in the AccountMenu, we can simplify this by just jusing accound.id directly.

rolznz replied:
Sorry, I misunderstood. I see this is for the Options version of TestConnection. I don't know what should happen in this case, probably it should navigate away from this page if the user selects a different account.

I am now thinking it might be better to have these TestConnection changes in a separate PR and keep this one as just replacing the custom SWR code with the SWR library. I think your were correct in saying that it isn't that important since this is only used once in setup.

lujakob replied:
@rolznz so we keep as is for now?

rolznz replied:
Maybe, or we could revert the TestConnection changes for now and put them in a separate PR later, once this one is merged.

lujakob replied:
I really recommend not to revert back. We would need to bring back a lot of code in AccountContext.fetchAccountInfo, api.swr.getAccountInfo and getAccountsCache that I already removed. I mean I can revert, back it is quite a lot of code and for no real reason. Right now we have a lot simplified. I suggest to keep as is. What would be the reason to revert and reimplement later?

rolznz replied:
Sounds good, let's keep it as is then.

I just suggested it to make this PR more easier to be reviewed by other Alby team members.

rolznz commented on src/app/screens/Onboard/TestConnection/index.tsx:
```diff
@@ -17,7 +17,7 @@ export default function TestConnection() {
   const navigate = useNavigate();
 
-  async function handleEdit(event: React.MouseEvent<HTMLButtonElement>) {
+  async function handleEdit() {
```
Are there any changes required in this component? I see there are two different components for testing connection (onboarding vs adding an account from the options).

Please note: I do not fully understand the code in TestConnection/AccountContext. Is there someone who has more knowledge of the code who might be able to review this? @escapedcat

escapedcat replied:
This looks like a "cleanup only", right @lujakob ? It's not needed but can be done I guess. Could be done in a different PR top keep things scoped but not a blocker really.
@rolznz is right, we have two of these. One for "adding an accoiunt via welcome/onboarding" and another one for "adding a new account after the onboarding already happened an d an accoiunt exists", which is here:
https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/screens/Options/TestConnection/index.tsx#L34-L37

Does this help?

lujakob replied:
Yes that's a minor cleanup. Should I revert it, to be addressed at another point?

There are two TestConnection components. The "loadAccountInfo" handler is different however. I guess they are different.

lujakob commented:
@rolznz @escapedcat thanks for reviewing. I updated accordingly.

lujakob commented:
@rolznz do you get why the e2e test doesn't succeed? Seems like problem with loading a dependency?

rolznz commented:
> @rolznz do you get why the e2e test doesn't succeed? Seems like problem with loading a dependency?

Yes, I think it is related to some changes that were recently merged into master. Hopefully it will be fixed soon.

escapedcat commented:
> @rolznz do you get why the e2e test doesn't succeed? Seems like problem with loading a dependency?

Yeah, no worries, no your fault. Restarting these helps. Did that now.

rolznz commented:
Some improvements in this PR:

- Instant account switching for accounts you have already connected to
- When switching account, it shows correct account icon & name
- When switching to a new account, it will show a loading component (needs work, and is only shown in the popup right now)
- Account context is a lot simpler (extracted status )

I am wondering now if we should have a loading screen for switching to a new connector rather than showing some skeleton loaders (we are not sure the connector can be connected to or not). For switching back to a connector you already connected to, it's now instant (because it uses SWR) but we should still check if the connection succeeds, otherwise show the connection failed screen.

CC @lujakob @reneaaron

lujakob commented:
> Some improvements in this PR:
> 
> * Instant account switching for accounts you have already connected to
> * When switching account, it shows correct account icon & name
> * When switching to a new account, it will show a loading component (needs work, and is only shown in the popup right now)
> * Account context is a lot simpler (extracted status )
> 
> I am wondering now if we should have a loading screen for switching to a new connector rather than showing some skeleton loaders (we are not sure the connector can be connected to or not). For switching back to a connector you already connected to, it's now instant (because it uses SWR) but we should still check if the connection succeeds, otherwise show the connection failed screen.
> 
> CC @lujakob @reneaaron
> 
> I wonder if it makes more sense to show a loading screen instead of skeleton loaders (#2456)

This PR is already 5 months old and interferes with a few current ones (account loading, error cases). I think we should finish the current ones and then in case re-address this one. Otherwise it gets even more confusing as it already is (we have several different issues on the topic of account loading/switch, error handling). Additionally this PR has a lot of changes and also has probably quite a few merge conflicts with master due to its age.

rolznz commented:
@lujakob I already addressed the merge conflicts. I don't know if the existing loading screen with skeleton loaders is the best way to do it, and does not work for the first account load (only switching accounts).

But I see what you mean - let's leave this issue for now and keep the discussion in the other tickets.
